From 59292f22cb67c739c87e95d1601d10bd748fa15a Mon Sep 17 00:00:00 2001
From: Yiyu Si <1486864380@qq.com>
Date: Thu, 18 Aug 2022 23:34:42 +0800
Subject: [PATCH] Merge RestartCommand


diff --git a/src/main/java/org/spigotmc/RestartCommand.java b/src/main/java/org/spigotmc/RestartCommand.java
index f15fd9f37..42a5ba192 100644
--- a/src/main/java/org/spigotmc/RestartCommand.java
+++ b/src/main/java/org/spigotmc/RestartCommand.java
@@ -7,6 +7,11 @@ import net.minecraft.server.MinecraftServer;
 import org.bukkit.command.Command;
 import org.bukkit.command.CommandSender;
 
+//Akarin Start
+import io.akarin.api.internal.Akari;
+import io.akarin.server.core.AkarinGlobalConfig;
+//Akarin End
+
 public class RestartCommand extends Command
 {
 
@@ -37,6 +42,15 @@ public class RestartCommand extends Command
 
     public static void restart()
     {
+        
+        //Akarin Start
+        if (AkarinGlobalConfig.noResponseDoGC) {
+            Akari.logger.warn("Attempting to garbage collect, may takes a few seconds");
+            System.runFinalization();
+            System.gc();
+        }
+        //Akarin End
+
         restart( new File( SpigotConfig.restartScript ) );
     }
 
-- 
2.35.1.windows.2

