From 7993d3da17dc7ab3d7de6fc25dcf8e7f414fad84 Mon Sep 17 00:00:00 2001
From: Yiyu Si <1486864380@qq.com>
Date: Fri, 19 Aug 2022 15:11:42 +0800
Subject: [PATCH] Akarin ChunkIOExecutor


diff --git a/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java b/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
index f50d55c8e..03285cde9 100644
--- a/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
+++ b/src/main/java/org/bukkit/craftbukkit/chunkio/ChunkIOExecutor.java
@@ -7,10 +7,11 @@ import net.minecraft.server.ChunkRegionLoader;
 import net.minecraft.server.MCUtil; // Paper
 import net.minecraft.server.World;
 import org.bukkit.craftbukkit.util.AsynchronousExecutor;
+import io.akarin.server.core.AkarinGlobalConfig; //Akarin
 
 public class ChunkIOExecutor {
     static final int BASE_THREADS = PaperConfig.minChunkLoadThreads; // Paper
-    static final int PLAYERS_PER_THREAD = 50;
+    static int PLAYERS_PER_THREAD = 50;
 
     private static final AsynchronousExecutor<QueuedChunk, Chunk, Runnable, RuntimeException> instance = new AsynchronousExecutor<QueuedChunk, Chunk, Runnable, RuntimeException>(new ChunkIOProvider(), BASE_THREADS);
 
@@ -28,7 +29,7 @@ public class ChunkIOExecutor {
     }
 
     public static void adjustPoolSize(int players) {
-        int size = Math.max(BASE_THREADS, (int) Math.ceil(players / PLAYERS_PER_THREAD));
+        int size = Math.max(BASE_THREADS, (int) Math.ceil(players / (PLAYERS_PER_THREAD = AkarinGlobalConfig.playersPerIOThread))); //Akarin
         instance.setActiveThreads(size);
     }
 
-- 
2.35.1.windows.2

